projects/extension/ai/semantic_catalog.py

```diff
@@ -0,0 +1,433 @@
+    parsed_config = get_parsed_config(plpy, catalog_name, config)
+
+    description = "None"
+
+    system_prompt = """
+    You are a SQL expert generates natural language descriptions of tables and views in a database.
+
+    The descriptions that you generate should be a concise single sentence.
+    """
+    message_content = f"""
+    Given the following table or view and some sample rows from it, generate a natural language description of it:
+
+    {get_obj_description(plpy, relation)}
+    {render_sample(plpy, relation)}
+    """
+    tools = [
+        {
+            "name": "generate_description",
+            "description": "Record description of table or view in well-formed JSON",
+            "input_schema": {
+                "type": "object",
+                "properties": {
+                    "description": {
+                        "type": "string",
+                        "description": "Description of the table or view",
+                    }
+                },
+            },
+        }
+    ]
+
+    provider = parsed_config.get("provider", None)
+    if provider == "anthropic":
+        model = parsed_config.get("model", "claude-3-5-sonnet-latest")
+        messages = [{"role": "user", "content": message_content}]
+
+        plan = plpy.prepare(
+            """
+                select ai.anthropic_generate(
+                    $1
+                    , $2
+                    , system_prompt => $3
+                    , tools => $4
+                    , tool_choice => '{{"type": "tool", "name": "generate_description"}}'::jsonb
+                )
+            """,
+            ["text", "jsonb", "text", "jsonb"],
+        )
+        result = plpy.execute(
+            plan, [model, json.dumps(messages), system_prompt, json.dumps(tools)]
+        )
+        response = json.loads(result[0]["anthropic_generate"])
+        description = response["content"][0]["input"]["description"]
+    elif provider == "ollama":
+        pass
+    elif provider == "openai":
+        model = parsed_config.get("model", "gpt-4o")
+        messages = [
+            {"role": "system", "content": system_prompt},
+            {"role": "user", "content": message_content},
+        ]
+        plan = plpy.prepare(
+            """
+                select ai.openai_chat_complete(
+                    $1
+                    , $2
+                    , tools => $3
+                    , tool_choice => '{{"type": "function", "function": {{"name": "generate_description"}}}}'
+                )
+            """,
+            ["text", "jsonb", "jsonb"],
+        )
+        result = plpy.execute(
+            plan, [model, json.dumps(messages), json.dumps(map_tools_to_openai(tools))]
+        )
+        response = json.loads(result[0]["openai_chat_complete"])
+        description = json.loads(
+            response["choices"][0]["message"]["tool_calls"][0]["function"]["arguments"]
+        )["description"]
+    elif provider == "cohere":
+        pass
```

Comment:
raise exception instead?

Reply:
Done in 27853696d476c88c7c8ae47fa003081e711dc8ed